### PR TITLE
Post code coverage comparison as a comment on PRs

### PR DESCRIPTION
**Pull request by andrewlock**
Adds a final stage to the "coverage" step in CI that grabs the current report, and the last coverage report from master, does a basic comparison of them, and posts it as a comment to the PR. A couple of examples below, and hopefully this PR should be updated with a comparison once tests complete.

> Note that this uses the _last successfully generated_ coverage report. That means that if the last `master` build failed due to flaky tests, the comparison will encompass more than just changes in this commit.

image
image

### CODE REVIEW

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #1704 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/master/6fa91cb147c882e13fcd1f58b97efa9da807c5bd) will **not change** line coverage
:heavy_check_mark: Merging #1704 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/master/6fa91cb147c882e13fcd1f58b97efa9da807c5bd) will **not change** branch coverage
:heavy_check_mark: Merging #1704 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/master/6fa91cb147c882e13fcd1f58b97efa9da807c5bd) will will **decrease** complexity by `1`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/master/6fa91cb147c882e13fcd1f58b97efa9da807c5bd) | #1704       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `8992` / `15199` | `8948` / `15190` |          |
| Lines %   | `0.59%`      | `0.59%`      |  `0.00%` :heavy_check_mark:  |
| Branches  | `3962` / `6594` | `3953` / `6594` |          |
| Branches %| `0.60%`      | `0.60%`      |  `0.00%` :heavy_check_mark:  |
| Complexity|   `7825`      | `7824`        |  `-1`  :heavy_check_mark:    |

View the full report for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/73546/artifacts?artifactName=Code%20Coverage%20Report_73546&api-version=6.0&%24format=zip)
* [For this PR #1704](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/73402/artifacts?artifactName=Code%20Coverage%20Report_73402&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/master/6fa91cb147c882e13fcd1f58b97efa9da807c5bd) | #1704       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `0.59%`      | `0.59%`       |  `0.00%` :heavy_check_mark:  |
| Branches %| `0.60%`      | `0.60%`       |  `0.00%` :heavy_check_mark:  |
| Complexity| `7825`      | `7824`       |  `-1` :heavy_check_mark:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.CallTarget.Handlers.Continuations.ValueTaskContinuationGenerator`3](https://github.com/DataDog/dd-trace-dotnet/tree/b5d1a395a8abe4b784db45be2cd801924bde8a06/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/Continuations/ValueTaskContinuationGenerator.cs) | `-0.72%` :no_entry: | `-0.67%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.CallTarget.Handlers.Continuations.ValueTaskContinuationGenerator`4](https://github.com/DataDog/dd-trace-dotnet/tree/b5d1a395a8abe4b784db45be2cd801924bde8a06/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/Continuations/ValueTaskContinuationGenerator`1.cs) | `-0.69%` :no_entry: | `-0.67%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.Agent.Transports.HttpStreamResponse](https://github.com/DataDog/dd-trace-dotnet/tree/b5d1a395a8abe4b784db45be2cd801924bde8a06/src/Datadog.Trace/Agent/Transports/HttpStreamResponse.cs) | `-0.10%` :no_entry: | `0.00%` :heavy_check_mark:   | `0` :heavy_check_mark: |
| [Datadog.Trace.HttpOverStreams.HttpHeaders](https://github.com/DataDog/dd-trace-dotnet/tree/b5d1a395a8abe4b784db45be2cd801924bde8a06/src/Datadog.Trace/HttpOverStreams/HttpHeaders.cs) | `-0.04%` :warning: | `-0.12%` :no_entry:   | `0` :heavy_check_mark: |


14 classes were removed from Datadog.Trace in #1704


View the full reports for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/73546/artifacts?artifactName=Code%20Coverage%20Report_73546&api-version=6.0&%24format=zip)
* [For this PR #1704](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/73402/artifacts?artifactName=Code%20Coverage%20Report_73402&api-version=6.0&%24format=zip)

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #1704 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/6fa91cb147c882e13fcd1f58b97efa9da807c5bd) will **not change** line coverage
:heavy_check_mark: Merging #1704 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/6fa91cb147c882e13fcd1f58b97efa9da807c5bd) will **not change** branch coverage
:heavy_check_mark: Merging #1704 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/6fa91cb147c882e13fcd1f58b97efa9da807c5bd) will **not change** complexity


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/6fa91cb147c882e13fcd1f58b97efa9da807c5bd) | #1704       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `8992` / `15199` | `8992` / `15199` |          |
| Lines %   | `0.59%`      | `0.59%`      |  `0.00%` :heavy_check_mark:  |
| Branches  | `3962` / `6594` | `3957` / `6594` |          |
| Branches %| `0.60%`      | `0.60%`      |  `0.00%` :heavy_check_mark:  |
| Complexity|   `7825`      | `7825`        |  `0`  :heavy_check_mark:    |

View the full report for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/73546/artifacts?artifactName=Code%20Coverage%20Report_73546&api-version=6.0&%24format=zip)
* [For this PR #1704](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/73610/artifacts?artifactName=Code%20Coverage%20Report_73610&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/6fa91cb147c882e13fcd1f58b97efa9da807c5bd) | #1704       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `0.59%`      | `0.59%`       |  `0.00%` :heavy_check_mark:  |
| Branches %| `0.60%`      | `0.60%`       |  `0.00%` :heavy_check_mark:  |
| Complexity| `7825`      | `7825`       |  `0` :heavy_check_mark:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.Logging.LogRateLimiter](https://github.com/DataDog/dd-trace-dotnet/tree/498150bef74c6589b86cce027d7cc668ff4e2783/src/Datadog.Trace/Logging/LogRateLimiter.cs) | `0.00%` :heavy_check_mark: | `-0.17%` :no_entry:   | `0` :heavy_check_mark: |


13 classes were removed from Datadog.Trace in #1704


View the full reports for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/73546/artifacts?artifactName=Code%20Coverage%20Report_73546&api-version=6.0&%24format=zip)
* [For this PR #1704](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/73610/artifacts?artifactName=Code%20Coverage%20Report_73610&api-version=6.0&%24format=zip)

**Comment by andrewlock** [comment]
Wait, what? 🤔 
image

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #1704 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/04538d71ef6af54f2ebea6c5b346808621b8e9e0) will **not change** line coverage
:heavy_check_mark: Merging #1704 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/04538d71ef6af54f2ebea6c5b346808621b8e9e0) will **not change** branch coverage
:no_entry: Merging #1704 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/04538d71ef6af54f2ebea6c5b346808621b8e9e0) will will **increase** complexity by `23`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/04538d71ef6af54f2ebea6c5b346808621b8e9e0) | #1704       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9352` / `15199` | `9392` / `15246` |          |
| Lines %   | `0.62%`      | `0.62%`      |  `0.00%` :heavy_check_mark:  |
| Branches  | `4138` / `6594` | `4149` / `6608` |          |
| Branches %| `0.63%`      | `0.63%`      |  `0.00%` :heavy_check_mark:  |
| Complexity|   `7825`      | `7848`        |  `23`  :no_entry:    |

View the full report for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/73832/artifacts?artifactName=Code%20Coverage%20Report_73832&api-version=6.0&%24format=zip)
* [For this PR #1704](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/73987/artifacts?artifactName=Code%20Coverage%20Report_73987&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/04538d71ef6af54f2ebea6c5b346808621b8e9e0) | #1704       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `0.62%`      | `0.62%`       |  `0.00%` :heavy_check_mark:  |
| Branches %| `0.63%`      | `0.63%`       |  `0.00%` :heavy_check_mark:  |
| Complexity| `7825`      | `7848`       |  `23` :no_entry:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.CallTarget.Handlers.Continuations.TaskContinuationGenerator`4](https://github.com/DataDog/dd-trace-dotnet/tree/95ef6a32a06cf333180093db04394f90a011a4ee/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/Continuations/TaskContinuationGenerator`1.cs) | `0.05%` :heavy_check_mark: | `0.09%` :heavy_check_mark:   | `0` :heavy_check_mark: |


13 classes were removed from Datadog.Trace in #1704


View the full reports for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/73832/artifacts?artifactName=Code%20Coverage%20Report_73832&api-version=6.0&%24format=zip)
* [For this PR #1704](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/73987/artifacts?artifactName=Code%20Coverage%20Report_73987&api-version=6.0&%24format=zip)

File: build/_build/Build.GitHub.cs [hunk 1]
```diff
@@ -377,90 +380,179 @@ partial class Build
...
+              var oldBuildId = oldArtifact.Name.Substring(fixedPrefix.Length);
+              var newBuildId = newArtifact.Name.Substring(fixedPrefix.Length);
+
+              var oldReportPath = oldReportdir / oldArtifact.Name / $"summary{oldBuildId}" / "Cobertura.xml";
+              var newReportPath = newReportdir / newArtifact.Name / $"summary{newBuildId}" / "Cobertura.xml";
+
+              var downloadOldLink = oldArtifact.Resource.DownloadUrl;
+              var downloadNewLink = newArtifact.Resource.DownloadUrl;
+
+              // var oldReportPath = RootDirectory / "Cobertura.old.xml";
```
Comment (zacharycmontoya) [inline build/_build/Build.GitHub.cs:427]: Debug comments?

File: build/_build/Covertura/CompareReports.cs [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+                    var changeSummary = new CoverturaReportComparison.ClassChanges
+                    {
+                        Name = oldClass.Name,
+                        Filename = oldClass.Filename,
+                        LineCoverageChange = decimal.Round(newClass.LineRate - oldClass.LineRate, decimals: 2),
+                        BranchCoverageChange = decimal.Round(newClass.BranchRate - oldClass.BranchRate, decimals: 2),
+                        ComplexityChange = newClass.Complexity - oldClass.Complexity,
+                    };
+                    // somewhat arbitrary
+                    changeSummary.IsSignificantChange = Math.Abs(changeSummary.LineCoverageChange) > 0.05m
```
Comment (zacharycmontoya) [inline build/_build/Covertura/CompareReports.cs:123]: I'm thinking the `0.05m` value can be abstracted into a constant so we can easily keep track of our decision criteria

File: build/_build/Covertura/CompareReports.cs [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+                                                     || Math.Abs(changeSummary.BranchCoverageChange) > 0.05m;
+
+                    changes.ClassChanges[classKvp.Key] = changeSummary;
+
+                }
+
+                changes.NewClasses =
+                    newPackage.Classes
+                             .Where(kvp => !newPackage.Classes.ContainsKey(kvp.Key))
```
Comment (zacharycmontoya) [inline build/_build/Covertura/CompareReports.cs:132]: I think there's a typo here:
```suggestion
                             .Where(kvp => !oldPackage.Classes.ContainsKey(kvp.Key))
```

**Review by zacharycmontoya — APPROVED** [review]
LGTM with a few small comments

File: build/_build/Build.GitHub.cs (hunk 1 above)
Comment (andrewlock) [reply build/_build/Build.GitHub.cs:427]: Ah, yes and no - I was using this to compare two _local_ coverage files. That might be useful enough to separate out into a dedicated target 🤔

Comment (andrewlock) [reply build/_build/Build.GitHub.cs]: Fixed in https://github.com/DataDog/dd-trace-dotnet/pull/1704/commits/8a7db0e5b58679872b803b47c25434462ba93783

File: build/_build/Covertura/CompareReports.cs (hunk 2 above)
Comment (andrewlock) [reply build/_build/Covertura/CompareReports.cs]: Fixed in https://github.com/DataDog/dd-trace-dotnet/pull/1704/commits/8a7db0e5b58679872b803b47c25434462ba93783

File: build/_build/Covertura/CompareReports.cs (hunk 3 above)
Comment (andrewlock) [reply build/_build/Covertura/CompareReports.cs]: Fixed in https://github.com/DataDog/dd-trace-dotnet/pull/1704/commits/8a7db0e5b58679872b803b47c25434462ba93783

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #1704 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/2ce5c96ba81f7552709a3688ccf4e15d245331c5) will **not change** line coverage
:heavy_check_mark: Merging #1704 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/2ce5c96ba81f7552709a3688ccf4e15d245331c5) will **not change** branch coverage
:heavy_check_mark: Merging #1704 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/2ce5c96ba81f7552709a3688ccf4e15d245331c5) will **not change** complexity


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/2ce5c96ba81f7552709a3688ccf4e15d245331c5) | #1704       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9254` / `14669` | `9256` / `14669` |          |
| Lines %   | `0.63%`      | `0.63%`      |  `0.00%` :heavy_check_mark:  |
| Branches  | `4119` / `6394` | `4120` / `6394` |          |
| Branches %| `0.64%`      | `0.64%`      |  `0.00%` :heavy_check_mark:  |
| Complexity|   `7544`      | `7544`        |  `0`  :heavy_check_mark:    |

View the full report for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/74696/artifacts?artifactName=Code%20Coverage%20Report_74696&api-version=6.0&%24format=zip)
* [For this PR #1704](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/74777/artifacts?artifactName=Code%20Coverage%20Report_74777&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/2ce5c96ba81f7552709a3688ccf4e15d245331c5) | #1704       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `0.63%`      | `0.63%`       |  `0.00%` :heavy_check_mark:  |
| Branches %| `0.64%`      | `0.64%`       |  `0.00%` :heavy_check_mark:  |
| Complexity| `7544`      | `7544`       |  `0` :heavy_check_mark:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.CallTarget.Handlers.Continuations.TaskContinuationGenerator`4](https://github.com/DataDog/dd-trace-dotnet/tree/b7dc0090c72c93dd2445d45a8aa926c250d6aefd/tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/Continuations/TaskContinuationGenerator`1.cs) | `-0.05%` :no_entry: | `-0.09%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.RuntimeMetrics.RuntimeMetricsWriter](https://github.com/DataDog/dd-trace-dotnet/tree/b7dc0090c72c93dd2445d45a8aa926c250d6aefd/tracer/src/Datadog.Trace/RuntimeMetrics/RuntimeMetricsWriter.cs) | `0.00%` :heavy_check_mark: | `-0.08%` :no_entry:   | `0` :heavy_check_mark: |


View the full reports for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/74696/artifacts?artifactName=Code%20Coverage%20Report_74696&api-version=6.0&%24format=zip)
* [For this PR #1704](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/74777/artifacts?artifactName=Code%20Coverage%20Report_74777&api-version=6.0&%24format=zip)

**Review by kevingosse — APPROVED** [review]
I didn't realize you could give the authentication token to Nuke to do this kind of thing (I mean, it's obvious in hindsight 😅 ). I guess we could do something similar to post the results of the throughput tests

**Comment by tonyredondo** [comment]
> I didn't realize you could give the authentication token to Nuke to do this kind of thing (I mean, it's obvious in hindsight 😅 ). I guess we could do something similar to post the results of the throughput tests

Haha I was thinking the same... also for benchmarks, to spot regressions. 😄

**Review by tonyredondo — APPROVED** [review]

**Pull request merged**